diff --git a/L1Trigger/L1TMuon/interface/deprecate/GeometryTranslator.h b/L1Trigger/L1TMuon/interface/deprecate/GeometryTranslator.h
index 8023aca..e0a15f9 100644
--- a/L1Trigger/L1TMuon/interface/deprecate/GeometryTranslator.h
+++ b/L1Trigger/L1TMuon/interface/deprecate/GeometryTranslator.h
@@ -42,6 +42,8 @@ namespace L1TMuon{
     double calculateGlobalPhi(const TriggerPrimitive&) const;
     double calculateBendAngle(const TriggerPrimitive&) const;    
 
+    GlobalPoint getGlobalPoint(const TriggerPrimitive&) const;
+
     void checkAndUpdateGeometry(const edm::EventSetup&);
 
   private:
diff --git a/L1Trigger/L1TMuon/src/GeometryTranslator.cc b/L1Trigger/L1TMuon/src/GeometryTranslator.cc
index d9290c1..aa0212c 100644
--- a/L1Trigger/L1TMuon/src/GeometryTranslator.cc
+++ b/L1Trigger/L1TMuon/src/GeometryTranslator.cc
@@ -78,6 +78,25 @@ GeometryTranslator::calculateBendAngle(const TriggerPrimitive& tp) const {
   }
 }
 
+GlobalPoint
+GeometryTranslator::getGlobalPoint(const TriggerPrimitive& tp) const {
+  switch(tp.subsystem()) {
+  case TriggerPrimitive::kDT:
+    return calcDTSpecificPoint(tp);
+    break;
+  case TriggerPrimitive::kCSC:
+    return getCSCSpecificPoint(tp);
+    break;
+  case TriggerPrimitive::kRPC:
+    return getRPCSpecificPoint(tp);
+    break;
+  default:
+    const GlobalPoint ret(GlobalPoint::Polar(0.0, 0.0, std::nan("Invalid TP type!")));
+    return ret;
+    break;
+  }
+}
+
 void GeometryTranslator::checkAndUpdateGeometry(const edm::EventSetup& es) {
   const MuonGeometryRecord& geom = es.get<MuonGeometryRecord>();
   unsigned long long geomid = geom.cacheIdentifier();
